File: backend/app/models/analysis.py
from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from app.models.findings import Finding
from app.models.transcript import TranscriptResult


class AnalysisStatus(StrEnum):
    QUEUED = "QUEUED"
    VALIDATING = "VALIDATING"
    EXTRACTING_METADATA = "EXTRACTING_METADATA"
    DETECTING_SILENCE = "DETECTING_SILENCE"
    TRANSCRIBING = "TRANSCRIBING"
    DETECTING_SCENES = "DETECTING_SCENES"
    BUILDING_REPORT = "BUILDING_REPORT"
    COMPLETED = "COMPLETED"
    PARTIAL_SUCCESS = "PARTIAL_SUCCESS"
    FAILED = "FAILED"


class VideoMetadata(BaseModel):
    fileName: str
    fileSizeBytes: int
    duration: float
    width: int
    height: int
    aspectRatio: float
    fps: float | None = None
    videoCodec: str | None = None
    audioCodec: str | None = None
    hasAudio: bool
    averageBitrate: int | None = None


class Scene(BaseModel):
    index: int
    start: float
    end: float
    duration: float


class Summary(BaseModel):
    readinessScore: int
    grade: str
    totalFindings: int
    highCount: int
    mediumCount: int
    lowCount: int
    infoCount: int


class AnalysisResult(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    analysisId: str
    status: AnalysisStatus
    createdAt: datetime
    completedAt: datetime | None = None
    currentStep: AnalysisStatus | None = None
    metadata: VideoMetadata | None = None
    summary: Summary | None = None
    findings: list[Finding] = Field(default_factory=list)
    transcript: TranscriptResult | None = None
    scenes: list[Scene] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    error: str | None = None


def utc_now() -> datetime:
    return datetime.now(UTC)

File: frontend/src/types/analysis.ts
export type AnalysisStatus =
  | "QUEUED"
  | "VALIDATING"
  | "EXTRACTING_METADATA"
  | "DETECTING_SILENCE"
  | "TRANSCRIBING"
  | "DETECTING_SCENES"
  | "BUILDING_REPORT"
  | "COMPLETED"
  | "PARTIAL_SUCCESS"
  | "FAILED";

export type Severity = "INFO" | "LOW" | "MEDIUM" | "HIGH";

export interface VideoMetadata {
  fileName: string;
  fileSizeBytes: number;
  duration: number;
  width: number;
  height: number;
  aspectRatio: number;
  fps: number | null;
  videoCodec: string | null;
  audioCodec: string | null;
  hasAudio: boolean;
  averageBitrate: number | null;
}

export interface Finding {
  id: string;
  type: string;
  severity: Severity;
  start: number;
  end: number;
  duration: number;
  title: string;
  reason: string;
  suggestion: string;
  source: string;
  metadata: Record<string, unknown>;
}

export interface TranscriptSegment {
  id: number;
  start: number;
  end: number;
  text: string;
  words: Array<{ start: number; end: number; word: string }>;
}

export interface TranscriptResult {
  language: string | null;
  languageProbability: number | null;
  duration: number | null;
  segments: TranscriptSegment[];
  error: string | null;
}

export interface Scene {
  index: number;
  start: number;
  end: number;
  duration: number;
}

export interface Summary {
  readinessScore: number;
  grade: string;
  totalFindings: number;
  highCount: number;
  mediumCount: number;
  lowCount: number;
  infoCount: number;
}

export interface AnalysisResult {
  analysisId: string;
  status: AnalysisStatus;
  createdAt: string;
  completedAt: string | null;
  currentStep: AnalysisStatus | null;
  metadata: VideoMetadata | null;
  summary: Summary | null;
  findings: Finding[];
  transcript: TranscriptResult | null;
  scenes: Scene[];
  warnings: string[];
  error: string | null;
}

